pkg/adaptation/os/adapter.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package os
+
+import (
+	goos "os"
+)
+
+type FileMode = goos.FileMode
+
+var Args = goos.Args
+
+// Adapter serves as a shim between between callers of standard os.* APIs
+// and the functions themselves.  The default behavior is simply to dispatch
+// the call to the corresponding os.* function.
+//
+// If the field associated with that function is non-nil, then the adapter will
+// dispatch to that function instead.
+type Adapter struct {
```

Comment:
Yep, that's definitely an option, and I've seen something like that done in the past.  When we'd submit jobs for automated tests, we'd occasionally get unexpected errors (e.g., something would misbehave and fill up the container's filesystem).

That also changes the validation portion of the tests from semi-simple asserts into "stat this file and make sure it exists" and "read this file and make sure it has the expected content".

In an ideal case, I would have written UTs for the mocks to try to avoid a scenario where some misbehavior there would cause unexpected results.  Here I didn't because they were fairly simple, and I didn't want to go too far down the rabbit hole given the time constraints on the exercise.

If you'd like, I can rework this to use the temp dir approach -- just let me know.

Reply:
> That also changes the validation portion of the tests from semi-simple asserts into "stat this file and make sure it exists" and "read this file and make sure it has the expected content".

This is one of the greatest benefits of the change, IMO - we are not doing asserts on special structs anymore, just reading files (which is what the program is supposed to output).

> If you'd like, I can rework this to use the temp dir approach -- just let me know.

I wanted to hear your thoughts more than anything else. We can keep as-is.